apps/web/app/blog/posts/mcp-sampling-support/page.mdx

```diff
@@ -0,0 +1,88 @@
+
+In the following example we use an `emojify_via_llm` MCP tool which uses Sampling to turn a string into emojis. Within the dropdown you can see the messages the tool sent to Tambo and the responses it received.
+
+<ImageZoom>
+  <img
+    src="/assets/blog/mcp-sampling-support/sampling-expand.gif"
+    alt="Sampling dropdown expansion"
+    className="max-w-lg mx-auto rounded-md"
+  />
+</ImageZoom>
+
+The title of the dropdown defaults to “finished additional work” but is customizable. This default is based on the opinion that the end app user might be confused by any specific terminology about MCP or Sampling. At the end of the day, the end user just needs to know that the AI is “doing something”. To customize this, update the `titleText` prop passed into the `SamplingSubThread` component within `message.tsx`.
+
+<ImageZoom>
+  <img
+    src="/assets/blog/mcp-sampling-support/custom-sampling-title.png"
+    alt="Sampling UI Screenshot"
+    className="max-w-md mx-auto rounded-md"
+  />
+</ImageZoom>
+
+## How to customize Tambo's Sampling UI
+
+Behind the scenes, messages sent and returned through sampling requests are added to the user’s conversation thread in Tambo with the same structure as other messages, except that they have a `parentMessageId` field pointing to the assistant message that triggered the toolcall resulting in the sampling request. This way we can clearly see them as “sub messages” of a normal conversation message.
+
+To build your own custom sampling UI, use Tambo’s [React SDK](https://github.com/tambo-ai/tambo) to get the list of messages for the current thread, and use the `parentMessageId` field on the messages to group and show sampling messages in any way you prefer.
+
+<ImageZoom>
+  <img
+    src="/assets/blog/mcp-sampling-support/custom-sampling-code.png"
+    alt="Sampling messages code"
+    className="max-w-full mx-auto rounded-md"
+  />
+</ImageZoom>
+
+## How to enable MCP Sampling in your app using Tambo
+
+If you already have an app with Tambo running, you’ll need to:
+
+- Upgrade your `tambo-ai/react` dependency to at least version `0.57.0` using `npm install @tambo-ai/react@0.57.0`
+- Upgrade your installed Tambo `Message` component using `npx tambo@latest upgrade message`
+
+If you haven’t integrated Tambo into your web app yet, new installations will automatically contain MCP Sampling functionality and UI. To get started, check out our [documentation](https://docs.tambo.co/getting-started/quickstart).
+
+## Up next for MCP Sampling support
+
+- Allow clientside sampling support for apps where the user is allowed to add their own mcp servers.
+- Enable user confirmation or rejection of sampling requests.
+- Show unique sampling requests separately within one toolcall.
```

Comment:
I think this is a subtle detail that is more confusing to call out than not. I had to read this 3 times before I knew what it was, and even then it was only because you and I talked on slack about it